cmd/format-xl.go

```diff
@@ -443,6 +446,81 @@ func checkFormatXLValues(formats []*formatXLV3) error {
 	return nil
 }
 
+// Get Deployment ID for the XL sets from format.json.
+// This need not be in quorum. Even if one of the format.json
+// file has this value, we assume it is valid.
+// If more than one format.json's have different id, it is considered a corrupt
+// backend format.
+func formatXLGetDeploymentID(refFormat *formatXLV3, formats []*formatXLV3) (string, error) {
+	var deploymentID string
+	for _, format := range formats {
+		if format == nil || format.ID == "" {
+			continue
+		}
+		if reflect.DeepEqual(format.XL.Sets, refFormat.XL.Sets) {
+			// Found an ID in one of the format.json file
+			// Set deploymentID for the first time.
+			if deploymentID == "" {
+				deploymentID = format.ID
+			} else if deploymentID != format.ID {
+				// DeploymentID found earlier doesn't match with the
+				// current format.json's ID.
+				return "", errCorruptedFormat
+			}
+		}
+	}
+	return deploymentID, nil
+}
+
+// formatXLFixDeploymentID - Add deployment id if it is not present.
+func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, refFormat *formatXLV3) (err error) {
+	// Acquire lock on format.json
+	mutex := newNSLock(globalIsDistXL)
+	formatLock := mutex.NewNSLock(minioMetaBucket, formatConfigFile)
+	if err = formatLock.GetLock(globalHealingTimeout); err != nil {
+		return err
+	}
+	defer formatLock.Unlock()
+	// Attempt to load all `format.json` from all disks.
+	formats, sErrs := loadFormatXLAll(storageDisks)
+	for i, sErr := range sErrs {
+		if _, ok := formatCriticalErrors[sErr]; ok {
+			return fmt.Errorf("Disk %s: %s", globalEndpoints[i], sErr)
+		}
+	}
+	refFormat.ID, err = formatXLGetDeploymentID(refFormat, formats)
+	if err != nil {
+		return err
+	}
```

Comment:
I think you can do:
```
if refFormat.ID == "" {

+	formats, sErrs := loadFormatXLAll(storageDisks)
+	for i, sErr := range sErrs {
+		if _, ok := formatCriticalErrors[sErr]; ok {
+			return fmt.Errorf("Disk %s: %s", globalEndpoints[i], sErr)
+		}
+	}
+	refFormat.ID, err = formatXLGetDeploymentID(refFormat, formats)
+	if err != nil {
+		return err
+	}

}

Reply:
i.e if refFormat.ID is already set, we dont have to re-read all the format jsons

Reply:
Currently, in this code flow, refFormat.ID will never be set as we do not read it in `getFormatXLInQuorum`. I can change that function to return `ID`.

Reply:
How about this change? I am not reading the format.json from the other disks again.
```
--- a/cmd/format-xl.go
+++ b/cmd/format-xl.go
@@ -473,7 +473,7 @@ func formatXLGetDeploymentID(refFormat *formatXLV3, formats []*formatXLV3) (stri
 }
 
 // formatXLFixDeploymentID - Add deployment id if it is not present.
-func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, refFormat *formatXLV3) (err error) {
+func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, refFormat *formatXLV3, formats []*formatXLV3) (err error) {
 	// Acquire lock on format.json
 	mutex := newNSLock(globalIsDistXL)
 	formatLock := mutex.NewNSLock(minioMetaBucket, formatConfigFile)
@@ -481,13 +481,7 @@ func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, ref
 		return err
 	}
 	defer formatLock.Unlock()
-	// Attempt to load all `format.json` from all disks.
-	formats, sErrs := loadFormatXLAll(storageDisks)
-	for i, sErr := range sErrs {
-		if _, ok := formatCriticalErrors[sErr]; ok {
-			return fmt.Errorf("Disk %s: %s", globalEndpoints[i], sErr)
-		}
-	}
+
 	refFormat.ID, err = formatXLGetDeploymentID(refFormat, formats)
 	if err != nil {
 		return err
@@ -507,7 +501,8 @@ func formatXLFixDeploymentID(ctx context.Context, storageDisks []StorageAPI, ref
 		return saveFormatXLAll(ctx, storageDisks, formats)
 	}
 
-	// Update if deploymentID is empty on a local disk
+	// If this node was down when the deploymentID was updated,
+	// then we make sure that we update the local disks with the deploymentID.
 	for index, format := range formats {
 		if format != nil && format.ID == "" && globalEndpoints[index].IsLocal {
 			format.ID = refFormat.ID
diff --git a/cmd/prepare-storage.go b/cmd/prepare-storage.go
index 54eb6ab3..e84231cf 100644
--- a/cmd/prepare-storage.go
+++ b/cmd/prepare-storage.go
@@ -186,7 +186,7 @@ func connectLoadInitFormats(firstDisk bool, endpoints EndpointList, setCount, dr
 		return nil, err
 	}
 
-	if err = formatXLFixDeploymentID(context.Background(), storageDisks, format); err != nil {
+	if err = formatXLFixDeploymentID(context.Background(), storageDisks, format, formatConfigs); err != nil {
 		return nil, err
 	}
```